### PR TITLE
[expo-module-scripts] Generate expo modules structure and mocks from swift

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

Ideally, we'd like to have first-class support for jest in expo-modules. We want module authors to be able to create jest mocks for native code easily, ideally automatically.

This is composed of 3 parts:

1. Reading structure of native code without running native apps (since it's slow).
2. Mapping that structure to mock files (e.g. packages/expo-crypto/mocks/ExpoCrypto.js) included in this PR
3. Loading these files into jest mocks when running our preset.

This PR is focused on 1 and 2. 
https://github.com/expo/expo/pull/24065/files#diff-74356e8b1c227906794f5ff7f366039cd4c05a8acc370b4fa2b81a5bffa54a26 is focused on 3.

# How

We're using sourcekitten to find and convert swift module definitions to a JSON structure.
The only thing we can't get access to in terms of types is a return type of a lambda.

# Results

Here is expo-crypto [getStructure](https://gist.github.com/aleqsio/1f065800f50cab5f36a06eb4e2d3d7a8) and [mock](https://gist.github.com/aleqsio/aaac9c967aa59f8b18d8bd8b48f9c1c6).

Here is[ expo-image structure](https://gist.github.com/aleqsio/c2575d255112714a57dfbf150d2dbb92).

# Test Plan

Tested manually.

# Checklist

TODO: Cleanup yarn.lock before merge
- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-modules-test-core/src/mockgen.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    case 'String':
+      return ts.factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword);
+    case 'Bool':
+      return ts.factory.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword);
+    case 'Int':
+    case 'Float':
+    case 'Double':
+      return ts.factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword);
+    default:
+      return ts.factory.createTypeReferenceNode(type);
```
Comment (tsapeta) [inline packages/expo-modules-test-core/src/mockgen.ts:46]: Is it like `any`? We support much more types on the Swift side, so we should extend the list of types here and have a fallback to `any` in other cases.

File: packages/expo-modules-test-core/src/mockgen.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+function getMockForModule(module: OutputModuleDefinition) {
+  return ([] as (ts.TypeAliasDeclaration | ts.FunctionDeclaration)[]).concat(
+    getMockedTypes(getTypesToMock(module)),
+    getMockedFunctions(module.functions),
+    getMockedFunctions(module.asyncFunctions, true)
+  );
+}
+
+export function printModules(modules: OutputModuleDefinition[]) {
```
Comment (tsapeta) [inline packages/expo-modules-test-core/src/mockgen.ts:143]: Name of the function is a bit confusing. I know that TS uses a "printer", but it's confusing if you don't know that. Maybe just `generateMocks`?

File: packages/expo-modules-test-core/src/getStructure.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+  file: FileType
+): OutputModuleDefinition {
+  const parsedDefinition = {
+    name: findNamedDefinitionsOfType('Name', moduleDefinition, file)?.[0]?.name,
+    functions: findNamedDefinitionsOfType('Function', moduleDefinition, file),
+    asyncFunctions: findNamedDefinitionsOfType('AsyncFunction', moduleDefinition, file),
+    events: findGroupedDefinitionsOfType('Events', moduleDefinition, file),
+    properties: findNamedDefinitionsOfType('Property', moduleDefinition, file),
+    props: omitViewFromClosureArguments(findNamedDefinitionsOfType('Prop', moduleDefinition, file)),
+    onCreate: findUnnamedDefinitionsOfType('OnCreate', moduleDefinition, file),
```
Comment (tsapeta) [inline packages/expo-modules-test-core/src/getStructure.ts:250]: Probably not necessary?

File: packages/expo-modules-test-core/src/getStructure.ts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+      parameters: d.types?.parameters?.filter((t, idx) => idx !== 0 && t.name !== 'view'),
+    },
+  }));
+}
+
+function parseModuleDefinition(
+  moduleDefinition: Structure[],
+  file: FileType
+): OutputModuleDefinition {
+  const parsedDefinition = {
```
Comment (tsapeta) [inline packages/expo-modules-test-core/src/getStructure.ts:243]: We will need to support classes at some point (it's fine to do it later)

File: packages/expo-modules-test-core/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { getAllExpoModulesInWorkingDirectory } from './getStructure';
+import { printModules } from './mockgen';
+
+const command = process.argv[2];
+
+if (command === 'mockgen') {
```
Comment (tsapeta) [inline packages/expo-modules-test-core/src/index.ts:6]: I like that you call it `mockgen`, but for the user something clear and simple might be better (e.g. `generate-mocks` or `generate-js-mocks`)

File: packages/expo-modules-test-core/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  const modules = getAllExpoModulesInWorkingDirectory();
+  printModules(modules);
+} else if (command === 'getStucture') {
```
Comment (tsapeta) [inline packages/expo-modules-test-core/src/index.ts:9]: kebab-case is more common for the CLI commands
```suggestion
} else if (command === 'get-mocks-structure') {
```

**Review by tsapeta — COMMENTED** [review]
Left some comments. Could you include in the PR description a screenshot of these commands so we can see what's their output?

**Comment by aleqsio** [comment]
@tsapeta 
Here is expo-crypto [getStructure](https://gist.github.com/aleqsio/1f065800f50cab5f36a06eb4e2d3d7a8) and [mock](https://gist.github.com/aleqsio/aaac9c967aa59f8b18d8bd8b48f9c1c6).

Here is[ expo-image structure](https://gist.github.com/aleqsio/c2575d255112714a57dfbf150d2dbb92).

**Review by tsapeta — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]

File: react-native-lab/react-native
Comment (tsapeta) [inline react-native-lab/react-native:1]: Hmm, something went wrong 🤔

**Pull request merged**